File: src/factories/MetadataFactory.ts
import ts from "typescript";

import { Metadata } from "../schemas/metadata/Metadata";
import { MetadataAliasType } from "../schemas/metadata/MetadataAliasType";
import { MetadataArrayType } from "../schemas/metadata/MetadataArrayType";
import { MetadataConstant } from "../schemas/metadata/MetadataConstant";
import { MetadataFunction } from "../schemas/metadata/MetadataFunction";
import { MetadataObject } from "../schemas/metadata/MetadataObject";
import { MetadataObjectType } from "../schemas/metadata/MetadataObjectType";
import { MetadataTupleType } from "../schemas/metadata/MetadataTupleType";
import { explore_metadata } from "./internal/metadata/explore_metadata";
import { iterate_metadata_collection } from "./internal/metadata/iterate_metadata_collection";
import { iterate_metadata_sort } from "./internal/metadata/iterate_metadata_sort";

import { ValidationPipe } from "../typings/ValidationPipe";

import { ExpressionFactory } from "./ExpressionFactory";
import { MetadataCollection } from "./MetadataCollection";

export namespace MetadataFactory {
  export type Validator = (meta: Metadata, explore: IExplore) => string[];

  export interface IProps {
    checker: ts.TypeChecker;
    transformer: ts.TransformationContext | undefined;
    options: IOptions;
    collection: MetadataCollection;
    type: ts.Type | null;
  }
  export interface IOptions {
    escape: boolean;
    constant: boolean;
    absorb: boolean;
    functional?: boolean;
    validate?: Validator;
    onError?: (node: ts.Node | undefined, message: string) => void;
  }
  export interface IExplore {
    top: boolean;
    object: MetadataObjectType | null;
    property: string | object | null;
    nested: null | MetadataAliasType | MetadataArrayType | MetadataTupleType;
    parameter: string | null;
    output: boolean;
    escaped: boolean;
    aliased: boolean;
  }

  export interface IError {
    name: string;
    explore: IExplore;
    messages: string[];
  }

  export const analyze = (props: IProps): ValidationPipe<Metadata, IError> => {
    const errors: IError[] = [];
    const metadata: Metadata = explore_metadata({
      ...props,
      errors,
      explore: {
        top: true,
        object: null,
        property: null,
        parameter: null,
        nested: null,
        aliased: false,
        escaped: false,
        output: false,
      },
      intersected: false,
    });
    iterate_metadata_collection({
      errors,
      collection: props.collection,
    });
    iterate_metadata_sort({
      collection: props.collection,
      metadata: metadata,
    });

    if (props.options.validate)
      errors.push(
        ...validate({
          transformer: props.transformer,
          options: props.options,
          functor: props.options.validate,
          metadata,
        }),
      );
    return errors.length
      ? {
          success: false,
          errors,
        }
      : {
          success: true,
          data: metadata,
        };
  };

  /** @internal */
  export const soleLiteral = (value: string): Metadata => {
    const meta: Metadata = Metadata.initialize();
    meta.constants.push(
      MetadataConstant.from({
        values: [
          {
            value,
            tags: [],
          },
        ],
        type: "string",
      }),
    );
    return meta;
  };

  export const validate = (props: {
    transformer?: ts.TransformationContext;
    options: IOptions;
    functor: Validator;
    metadata: Metadata;
  }): IError[] => {
    const visitor: IValidationVisitor = {
      functor: props.functor,
      errors: [],
      objects: new Set(),
      arrays: new Set(),
      tuples: new Set(),
      aliases: new Set(),
      functions: new Set(),
    };
    validateMeta({
      ...props,
      visitor,
      explore: {
        object: null,
        property: null,
        parameter: null,
        nested: null,
        top: true,
        aliased: false,
        escaped: false,
        output: false,
      },
    });
    return visitor.errors;
  };

  const validateMeta = (props: {
    options: IOptions;
    visitor: IValidationVisitor;
    metadata: Metadata;
    explore: IExplore;
  }) => {
    const result: string[] = [];
    for (const atomic of props.metadata.atomics)
      for (const row of atomic.tags)
        for (const tag of row.filter(
          (t) => t.validate !== undefined && t.predicate === undefined,
        ))
          try {
            tag.predicate = ExpressionFactory.transpile({
              script: tag.validate!,
            });
          } catch {
            result.push(
              `Unable to transpile type tag script: ${JSON.stringify(
                tag.validate,
              )}`,
            );
            tag.predicate = () => ts.factory.createTrue();
          }
    result.push(...props.visitor.functor(props.metadata, props.explore));
    if (result.length)
      props.visitor.errors.push({
        name: props.metadata.getName(),
        explore: { ...props.explore },
        messages: [...new Set(result)],
      });

    for (const alias of props.metadata.aliases)
      validateAlias({
        ...props,
        alias: alias.type,
      });
    for (const array of props.metadata.arrays)
      validateArray({
        ...props,
        array: array.type,
      });
    for (const tuple of props.metadata.tuples)
      validateTuple({
        ...props,
        tuple: tuple.type,
      });
    for (const object of props.metadata.objects)
      validateObject({
        ...props,
        object: object.type,
      });
    for (const func of props.metadata.functions)
      validateFunction({
        ...props,
        function: func,
      });
    for (const set of props.metadata.sets)
      validateMeta({
        ...props,
        metadata: set.value,
      });
    for (const map of props.metadata.maps) {
      validateMeta({
        ...props,
        metadata: map.key,
      });
      validateMeta({
        ...props,
        metadata: map.value,
      });
    }

    if (props.options.escape === true && props.metadata.escaped !== null)
      validateMeta({
        ...props,
        metadata: props.metadata.escaped.returns,
        explore: {
          ...props.explore,
          escaped: true,
        },
      });
  };

  const validateAlias = (props: {
    transformer?: ts.TransformationContext;
    options: IOptions;
    visitor: IValidationVisitor;
    alias: MetadataAliasType;
    explore: IExplore;
  }) => {
    if (props.visitor.aliases.has(props.alias)) return;
    props.visitor.aliases.add(props.alias);

    validateMeta({
      ...props,
      metadata: props.alias.value,
      explore: {
        ...props.explore,
        nested: props.alias,
        aliased: true,
      },
    });
  };

  const validateArray = (props: {
    transformer?: ts.TransformationContext;
    options: IOptions;
    visitor: IValidationVisitor;
    array: MetadataArrayType;
    explore: IExplore;
  }) => {
    if (props.visitor.arrays.has(props.array)) return;
    props.visitor.arrays.add(props.array);

    validateMeta({
      ...props,
      metadata: props.array.value,
      explore: {
        ...props.explore,
        nested: props.array,
        top: false,
      },
    });
  };

  const validateTuple = (props: {
    transformer?: ts.TransformationContext;
    options: IOptions;
    visitor: IValidationVisitor;
    tuple: MetadataTupleType;
    explore: IExplore;
  }) => {
    if (props.visitor.tuples.has(props.tuple)) return;
    props.visitor.tuples.add(props.tuple);

    for (const elem of props.tuple.elements)
      validateMeta({
        ...props,
        metadata: elem,
        explore: {
          ...props.explore,
          nested: props.tuple,
          top: false,
        },
      });
  };

  const validateObject = (props: {
    transformer?: ts.TransformationContext;
    options: IOptions;
    visitor: IValidationVisitor;
    object: MetadataObjectType;
  }) => {
    if (props.visitor.objects.has(props.object)) return;
    props.visitor.objects.add(props.object);

    if (props.options.validate) {
      const explore: IExplore = {
        object: props.object,
        top: false,
        property: null,
        parameter: null,
        nested: null,
        aliased: false,
        escaped: false,
        output: false,
      };
      const errors: string[] = props.options.validate(
        Metadata.create({
          ...Metadata.initialize(),
          objects: [
            MetadataObject.create({
              type: props.object,
              tags: [],
            }),
          ],
        }),
        explore,
      );
      if (errors.length)
        props.visitor.errors.push({
          name: props.object.name,
          explore,
          messages: [...new Set(errors)],
        });
    }

    for (const property of props.object.properties)
      validateMeta({
        ...props,
        metadata: property.value,
        explore: {
          object: props.object,
          property: property.key.isSoleLiteral()
            ? property.key.getSoleLiteral()!
            : {},
          parameter: null,
          nested: null,
          top: false,
          aliased: false,
          escaped: false,
          output: false,
        },
      });
  };

  const validateFunction = (props: {
    transformer?: ts.TransformationContext;
    options: IOptions;
    visitor: IValidationVisitor;
    function: MetadataFunction;
    explore: IExplore;
  }) => {
    if (props.visitor.functions.has(props.function)) return;
    props.visitor.functions.add(props.function);

    for (const param of props.function.parameters)
      validateMeta({
        ...props,
        metadata: param.type,
        explore: {
          ...props.explore,
          parameter: param.name,
          nested: null,
          top: false,
          output: false,
        },
      });
    if (props.function.output)
      validateMeta({
        ...props,
        metadata: props.function.output,
        explore: {
          ...props.explore,
          parameter: null,
          nested: null,
          top: false,
          output: true,
        },
      });
  };

  interface IValidationVisitor {
    functor: Validator;
    errors: IError[];
    objects: Set<MetadataObjectType>;
    arrays: Set<MetadataArrayType>;
    tuples: Set<MetadataTupleType>;
    aliases: Set<MetadataAliasType>;
    functions: Set<MetadataFunction>;
  }
}

File: packages/typia/native/core/factories/MetadataFactory.go
package factories

import (
  nativechecker "github.com/microsoft/typescript-go/shim/checker"
  nativemetadata "github.com/samchon/typia/packages/typia/native/core/factories/internal/metadata"
  schemametadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
)

type metadataFactoryNamespace struct{}

var MetadataFactory = metadataFactoryNamespace{}

func init() {
  nativemetadata.MetadataTypeTagAnalyzer = func(props struct {
    Errors  *[]nativemetadata.MetadataFactory_IError
    Type    string
    Objects []*schemametadata.MetadataObjectType
    Explore nativemetadata.MetadataFactory_IExplore
  }) []schemametadata.IMetadataTypeTag {
    return MetadataTypeTagFactory.Analyze(struct {
      Errors  *[]MetadataFactory_IError
      Type    string
      Objects []*schemametadata.MetadataObjectType
      Explore MetadataFactory_IExplore
    }{
      Errors:  props.Errors,
      Type:    props.Type,
      Objects: props.Objects,
      Explore: props.Explore,
    })
  }
}

type MetadataFactory_Validator = func(props struct {
  Metadata *schemametadata.MetadataSchema
  Explore  MetadataFactory_IExplore
  Top      *schemametadata.MetadataSchema
}) []string

type MetadataFactory_IOptions = nativemetadata.MetadataFactory_IOptions
type MetadataFactory_IExplore = nativemetadata.MetadataFactory_IExplore
type MetadataFactory_IError = nativemetadata.MetadataFactory_IError

type MetadataFactory_IProps struct {
  Checker     *nativechecker.Checker
  Transformer any
  Options     MetadataFactory_IOptions
  Components  *schemametadata.MetadataCollection
  Type        *nativechecker.Type
}

type MetadataFactory_ValidationPipe struct {
  Success bool
  Data    *schemametadata.MetadataSchema
  Errors  []MetadataFactory_IError
}

func (metadataFactoryNamespace) Analyze(props MetadataFactory_IProps) MetadataFactory_ValidationPipe {
  errors := []MetadataFactory_IError{}
  metadata := nativemetadata.Explore_metadata(nativemetadata.Explore_metadata_IProps{
    Options:    props.Options,
    Checker:    props.Checker,
    Components: props.Components,
    Errors:     &errors,
    Type:       props.Type,
    Explore: MetadataFactory_IExplore{
      Top:       true,
      Object:    nil,
      Property:  nil,
      Parameter: nil,
      Nested:    nil,
      Aliased:   false,
      Escaped:   false,
      Output:    false,
    },
    Intersected: false,
  })
  nativemetadata.Iterate_metadata_collection(struct {
    Errors     *[]MetadataFactory_IError
    Collection *schemametadata.MetadataCollection
  }{
    Errors:     &errors,
    Collection: props.Components,
  })
  nativemetadata.Iterate_metadata_sort(struct {
    Collection *schemametadata.MetadataCollection
    Metadata   *schemametadata.MetadataSchema
  }{
    Collection: props.Components,
    Metadata:   metadata,
  })

  if props.Options.Validate != nil {
    errors = append(errors, MetadataFactory.Validate(struct {
      Transformer any
      Options     MetadataFactory_IOptions
      Functor     MetadataFactory_Validator
      Metadata    *schemametadata.MetadataSchema
    }{
      Transformer: props.Transformer,
      Options:     props.Options,
      Functor:     props.Options.Validate,
      Metadata:    metadata,
    })...)
  }
  if len(errors) != 0 {
    return MetadataFactory_ValidationPipe{
      Success: false,
      Errors:  errors,
    }
  }
  return MetadataFactory_ValidationPipe{
    Success: true,
    Data:    metadata,
  }
}

func (metadataFactoryNamespace) SoleLiteral(value string) *schemametadata.MetadataSchema {
  meta := schemametadata.MetadataSchema_initialize()
  meta.Constants = append(meta.Constants, schemametadata.MetadataConstant_create(schemametadata.MetadataConstant{
    Values: []*schemametadata.MetadataConstantValue{
      schemametadata.MetadataConstantValue_create(schemametadata.MetadataConstantValue{
        Value: value,
        Tags:  [][]schemametadata.IMetadataTypeTag{},
      }),
    },
    Type: "string",
  }))
  return meta
}

func (metadataFactoryNamespace) Validate(props struct {
  Transformer any
  Options     MetadataFactory_IOptions
  Functor     MetadataFactory_Validator
  Metadata    *schemametadata.MetadataSchema
}) []MetadataFactory_IError {
  visitor := &metadataFactory_IValidationVisitor{
    Functor:   props.Functor,
    Errors:    []MetadataFactory_IError{},
    Objects:   map[*schemametadata.MetadataObjectType]bool{},
    Arrays:    map[*schemametadata.MetadataArrayType]bool{},
    Tuples:    map[*schemametadata.MetadataTupleType]bool{},
    Aliases:   map[*schemametadata.MetadataAliasType]bool{},
    Functions: map[*schemametadata.MetadataFunction]bool{},
  }
  MetadataFactory.validateMeta(struct {
    Options  MetadataFactory_IOptions
    Visitor  *metadataFactory_IValidationVisitor
    Metadata *schemametadata.MetadataSchema
    Top      *schemametadata.MetadataSchema
    Explore  MetadataFactory_IExplore
  }{
    Options:  props.Options,
    Visitor:  visitor,
    Metadata: props.Metadata,
    Top:      props.Metadata,
    Explore: MetadataFactory_IExplore{
      Object:    nil,
      Property:  nil,
      Parameter: nil,
      Nested:    nil,
      Top:       true,
      Aliased:   false,
      Escaped:   false,
      Output:    false,
    },
  })
  return visitor.Errors
}

func (metadataFactoryNamespace) validateMeta(props struct {
  Options  MetadataFactory_IOptions
  Visitor  *metadataFactory_IValidationVisitor
  Metadata *schemametadata.MetadataSchema
  Top      *schemametadata.MetadataSchema
  Explore  MetadataFactory_IExplore
}) {
  result := []string{}
  if props.Visitor.Functor != nil {
    result = append(result, props.Visitor.Functor(struct {
      Metadata *schemametadata.MetadataSchema
      Explore  MetadataFactory_IExplore
      Top      *schemametadata.MetadataSchema
    }{
      Metadata: props.Metadata,
      Explore:  props.Explore,
      Top:      props.Top,
    })...)
  }
  if len(result) != 0 {
    props.Visitor.Errors = append(props.Visitor.Errors, MetadataFactory_IError{
      Name:     props.Metadata.GetName(),
      Explore:  props.Explore,
      Messages: metadataFactory_unique(result),
    })
  }

  for _, alias := range props.Metadata.Aliases {
    MetadataFactory.validateAlias(struct {
      Options MetadataFactory_IOptions
      Visitor *metadataFactory_IValidationVisitor
      Alias   *schemametadata.MetadataAliasType
      Explore MetadataFactory_IExplore
      Top     *schemametadata.MetadataSchema
    }{
      Options: props.Options,
      Visitor: props.Visitor,
      Alias:   alias.Type,
      Explore: props.Explore,
      Top:     props.Top,
    })
  }
  for _, array := range props.Metadata.Arrays {
    MetadataFactory.validateArray(struct {
      Options MetadataFactory_IOptions
      Visitor *metadataFactory_IValidationVisitor
      Array   *schemametadata.MetadataArrayType
      Explore MetadataFactory_IExplore
      Top     *schemametadata.MetadataSchema
    }{
      Options: props.Options,
      Visitor: props.Visitor,
      Array:   array.Type,
      Explore: props.Explore,
      Top:     props.Top,
    })
  }
  for _, tuple := range props.Metadata.Tuples {
    MetadataFactory.validateTuple(struct {
      Options MetadataFactory_IOptions
      Visitor *metadataFactory_IValidationVisitor
      Tuple   *schemametadata.MetadataTupleType
      Explore MetadataFactory_IExplore
      Top     *schemametadata.MetadataSchema
    }{
      Options: props.Options,
      Visitor: props.Visitor,
      Tuple:   tuple.Type,
      Explore: props.Explore,
      Top:     props.Top,
    })
  }
  for _, object := range props.Metadata.Objects {
    MetadataFactory.validateObject(struct {
      Options MetadataFactory_IOptions
      Visitor *metadataFactory_IValidationVisitor
      Object  *schemametadata.MetadataObjectType
      Top     *schemametadata.MetadataSchema
    }{
      Options: props.Options,
      Visitor: props.Visitor,
      Object:  object.Type,
      Top:     props.Top,
    })
  }
  for _, function := range props.Metadata.Functions {
    MetadataFactory.validateFunction(struct {
      Options  MetadataFactory_IOptions
      Visitor  *metadataFactory_IValidationVisitor
      Function *schemametadata.MetadataFunction
      Explore  MetadataFactory_IExplore
      Top      *schemametadata.MetadataSchema
    }{
      Options:  props.Options,
      Visitor:  props.Visitor,
      Function: function,
      Explore:  props.Explore,
      Top:      props.Top,
    })
  }
  for _, set := range props.Metadata.Sets {
    MetadataFactory.validateMeta(struct {
      Options  MetadataFactory_IOptions
      Visitor  *metadataFactory_IValidationVisitor
      Metadata *schemametadata.MetadataSchema
      Top      *schemametadata.MetadataSchema
      Explore  MetadataFactory_IExplore
    }{
      Options:  props.Options,
      Visitor:  props.Visitor,
      Metadata: set.Value,
      Top:      props.Top,
      Explore:  props.Explore,
    })
  }
  for _, m := range props.Metadata.Maps {
    MetadataFactory.validateMeta(struct {
      Options  MetadataFactory_IOptions
      Visitor  *metadataFactory_IValidationVisitor
      Metadata *schemametadata.MetadataSchema
      Top      *schemametadata.MetadataSchema
      Explore  MetadataFactory_IExplore
    }{Options: props.Options, Visitor: props.Visitor, Metadata: m.Key, Top: props.Top, Explore: props.Explore})
    MetadataFactory.validateMeta(struct {
      Options  MetadataFactory_IOptions
      Visitor  *metadataFactory_IValidationVisitor
      Metadata *schemametadata.MetadataSchema
      Top      *schemametadata.MetadataSchema
      Explore  MetadataFactory_IExplore
    }{Options: props.Options, Visitor: props.Visitor, Metadata: m.Value, Top: props.Top, Explore: props.Explore})
  }
  if props.Options.Escape == true && props.Metadata.Escaped != nil {
    explore := props.Explore
    explore.Escaped = true
    MetadataFactory.validateMeta(struct {
      Options  MetadataFactory_IOptions
      Visitor  *metadataFactory_IValidationVisitor
      Metadata *schemametadata.MetadataSchema
      Top      *schemametadata.MetadataSchema
      Explore  MetadataFactory_IExplore
    }{Options: props.Options, Visitor: props.Visitor, Metadata: props.Metadata.Escaped.Returns, Top: props.Top, Explore: explore})
  }
}

func (metadataFactoryNamespace) validateAlias(props struct {
  Options MetadataFactory_IOptions
  Visitor *metadataFactory_IValidationVisitor
  Alias   *schemametadata.MetadataAliasType
  Explore MetadataFactory_IExplore
  Top     *schemametadata.MetadataSchema
}) {
  if props.Visitor.Aliases[props.Alias] {
    return
  }
  props.Visitor.Aliases[props.Alias] = true
  explore := props.Explore
  explore.Nested = props.Alias
  explore.Aliased = true
  MetadataFactory.validateMeta(struct {
    Options  MetadataFactory_IOptions
    Visitor  *metadataFactory_IValidationVisitor
    Metadata *schemametadata.MetadataSchema
    Top      *schemametadata.MetadataSchema
    Explore  MetadataFactory_IExplore
  }{Options: props.Options, Visitor: props.Visitor, Metadata: props.Alias.Value, Top: props.Top, Explore: explore})
}

func (metadataFactoryNamespace) validateArray(props struct {
  Options MetadataFactory_IOptions
  Visitor *metadataFactory_IValidationVisitor
  Array   *schemametadata.MetadataArrayType
  Explore MetadataFactory_IExplore
  Top     *schemametadata.MetadataSchema
}) {
  if props.Visitor.Arrays[props.Array] {
    return
  }
  props.Visitor.Arrays[props.Array] = true
  explore := props.Explore
  explore.Nested = props.Array
  explore.Top = false
  MetadataFactory.validateMeta(struct {
    Options  MetadataFactory_IOptions
    Visitor  *metadataFactory_IValidationVisitor
    Metadata *schemametadata.MetadataSchema
    Top      *schemametadata.MetadataSchema
    Explore  MetadataFactory_IExplore
  }{Options: props.Options, Visitor: props.Visitor, Metadata: props.Array.Value, Top: props.Top, Explore: explore})
}

func (metadataFactoryNamespace) validateTuple(props struct {
  Options MetadataFactory_IOptions
  Visitor *metadataFactory_IValidationVisitor
  Tuple   *schemametadata.MetadataTupleType
  Explore MetadataFactory_IExplore
  Top     *schemametadata.MetadataSchema
}) {
  if props.Visitor.Tuples[props.Tuple] {
    return
  }
  props.Visitor.Tuples[props.Tuple] = true
  for _, elem := range props.Tuple.Elements {
    explore := props.Explore
    explore.Nested = props.Tuple
    explore.Top = false
    MetadataFactory.validateMeta(struct {
      Options  MetadataFactory_IOptions
      Visitor  *metadataFactory_IValidationVisitor
      Metadata *schemametadata.MetadataSchema
      Top      *schemametadata.MetadataSchema
      Explore  MetadataFactory_IExplore
    }{Options: props.Options, Visitor: props.Visitor, Metadata: elem, Top: props.Top, Explore: explore})
  }
}

func (metadataFactoryNamespace) validateObject(props struct {
  Options MetadataFactory_IOptions
  Visitor *metadataFactory_IValidationVisitor
  Object  *schemametadata.MetadataObjectType
  Top     *schemametadata.MetadataSchema
}) {
  if props.Visitor.Objects[props.Object] {
    return
  }
  props.Visitor.Objects[props.Object] = true
  if props.Options.Validate != nil {
    explore := MetadataFactory_IExplore{
      Object:    props.Object,
      Top:       false,
      Property:  nil,
      Parameter: nil,
      Nested:    nil,
      Aliased:   false,
      Escaped:   false,
      Output:    false,
    }
    metadata := schemametadata.MetadataSchema_create(schemametadata.MetadataSchema{
      Any:       false,
      Nullable:  false,
      Required:  true,
      Optional:  false,
      Constants: []*schemametadata.MetadataConstant{},
      Atomics:   []*schemametadata.MetadataAtomic{},
      Templates: []*schemametadata.MetadataTemplate{},
      Arrays:    []*schemametadata.MetadataArray{},
      Tuples:    []*schemametadata.MetadataTuple{},
      Objects: []*schemametadata.MetadataObject{
        schemametadata.MetadataObject_create(schemametadata.MetadataObject{Type: props.Object, Tags: [][]schemametadata.IMetadataTypeTag{}}),
      },
      Aliases:   []*schemametadata.MetadataAlias{},
      Functions: []*schemametadata.MetadataFunction{},
      Rest:      nil,
      Natives:   []*schemametadata.MetadataNative{},
      Sets:      []*schemametadata.MetadataSet{},
      Maps:      []*schemametadata.MetadataMap{},
    })
    errors := props.Options.Validate(struct {
      Metadata *schemametadata.MetadataSchema
      Explore  MetadataFactory_IExplore
      Top      *schemametadata.MetadataSchema
    }{Metadata: metadata, Top: props.Top, Explore: explore})
    if len(errors) != 0 {
      props.Visitor.Errors = append(props.Visitor.Errors, MetadataFactory_IError{
        Name:     props.Object.Name,
        Explore:  explore,
        Messages: metadataFactory_unique(errors),
      })
    }
  }

  for _, property := range props.Object.Properties {
    var key any = struct{}{}
    if sole := property.Key.GetSoleLiteral(); sole != nil {
      key = *sole
    }
    explore := MetadataFactory_IExplore{
      Object:    props.Object,
      Property:  key,
      Parameter: nil,
      Nested:    nil,
      Top:       false,
      Aliased:   false,
      Escaped:   false,
      Output:    false,
    }
    MetadataFactory.validateMeta(struct {
      Options  MetadataFactory_IOptions
      Visitor  *metadataFactory_IValidationVisitor
      Metadata *schemametadata.MetadataSchema
      Top      *schemametadata.MetadataSchema
      Explore  MetadataFactory_IExplore
    }{Options: props.Options, Visitor: props.Visitor, Metadata: property.Value, Top: props.Top, Explore: explore})
  }
}

func (metadataFactoryNamespace) validateFunction(props struct {
  Options  MetadataFactory_IOptions
  Visitor  *metadataFactory_IValidationVisitor
  Function *schemametadata.MetadataFunction
  Explore  MetadataFactory_IExplore
  Top      *schemametadata.MetadataSchema
}) {
  if props.Visitor.Functions[props.Function] {
    return
  }
  props.Visitor.Functions[props.Function] = true
  for _, param := range props.Function.Parameters {
    explore := props.Explore
    explore.Parameter = param.Name
    explore.Nested = nil
    explore.Top = false
    explore.Output = false
    MetadataFactory.validateMeta(struct {
      Options  MetadataFactory_IOptions
      Visitor  *metadataFactory_IValidationVisitor
      Metadata *schemametadata.MetadataSchema
      Top      *schemametadata.MetadataSchema
      Explore  MetadataFactory_IExplore
    }{Options: props.Options, Visitor: props.Visitor, Metadata: param.Type, Top: props.Top, Explore: explore})
  }
  if props.Function.Output != nil {
    explore := props.Explore
    explore.Parameter = nil
    explore.Nested = nil
    explore.Top = false
    explore.Output = true
    MetadataFactory.validateMeta(struct {
      Options  MetadataFactory_IOptions
      Visitor  *metadataFactory_IValidationVisitor
      Metadata *schemametadata.MetadataSchema
      Top      *schemametadata.MetadataSchema
      Explore  MetadataFactory_IExplore
    }{Options: props.Options, Visitor: props.Visitor, Metadata: props.Function.Output, Top: props.Top, Explore: explore})
  }
}

type metadataFactory_IValidationVisitor struct {
  Functor   MetadataFactory_Validator
  Errors    []MetadataFactory_IError
  Objects   map[*schemametadata.MetadataObjectType]bool
  Arrays    map[*schemametadata.MetadataArrayType]bool
  Tuples    map[*schemametadata.MetadataTupleType]bool
  Aliases   map[*schemametadata.MetadataAliasType]bool
  Functions map[*schemametadata.MetadataFunction]bool
}

func metadataFactory_unique(input []string) []string {
  visited := map[string]bool{}
  output := []string{}
  for _, str := range input {
    if visited[str] {
      continue
    }
    visited[str] = true
    output = append(output, str)
  }
  return output
}
